### ISSUE TITLE
[Data] Parallelism in `flat_map`?

### ISSUE THREAD

**Issue by oceanusxiv** [ray-project/ray#45579]
### Description

Maybe I'm missing something, but how does achieve parallelism during a `flat_map` operation? Certain `flat_map` ops can generate multiple output rows from a single input row in a completely parallel fashion. 

As an example. Say one has a nested dataset. Where each row is a `list[struct]` of `num_structs` length, the desired output is an unrolled `num_rows*num_structs` rows of `struct`. In ray data it seems I would want to use `flat_map` with an unrolling function, but currently it seems I have to just do this in a single threaded fashion (e.g. a for loop)? What is a supported way to let each `flat_map` worker itself achieve parallelism within the function?

Side note: `tf.data.Dataset.interleave` seems to offer such an option, being able to parallelize both the read and map function itself.

**Comment by wingkitlee0** [comment]
Ray's flat_map should work out-of-the-box to parallelize the work for each row, e.g., each cpu processes on each row. Parallelism inside a UDF is possible and handled by the user.
Are you seeing low utilization or something?

**Comment by oceanusxiv** [comment]
@wingkitlee0 I am seeing low utilization yes. I tried using a ThreadPoolExecutor to achieve parallelism within the flat_map function, but that doesn't seem to be able to spawn more threads or speed up the flat_map.

**Comment by raulchen** [comment]
it could be because the input op doesn't produce enough number of blocks. you can manually override it with `read_parquet(..., override_num_blocks=N)`

**Comment by oceanusxiv** [comment]
@raulchen Yeah but the price of increasing number of blocks is drastically increased memory usage. Especially for reading binary files. Shouldn't it be possible, when one binary file contains many rows, to have a single reader on the binary blob, but multiple workers while flat mapping it?
